fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,106 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos')
+        .then(res=>res.json())
+        .then(users=>{
+          this.setState({
+            todos:users
+        })})
+        .catch(err=>console.error(err));
+  }
+
+  addTodo(newTodo) {
+    /**
+     * Use fetch to send a POST request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * add the created todo to the state array.
+     *
+     * TIP: Be sure to SERIALIZE the todo in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    fetch('/api/todos',{
+      method:'POST',
+      headers:{'Content-type':'application/json;charset=UTF-8'},
+      body:JSON.stringify(newTodo),
+    })
+        .then(res=>res.json())
+        .then(()=>{
+          this.state.todos.push(newTodo);

```

Comment:
You should never directly change state. Here you want to make a copy of `this.state.todos` using something like `.slice` then push the `todo` parameter into the copy of the todos array, then finally use the copy to set state.